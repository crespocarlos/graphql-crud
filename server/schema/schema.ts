import { GraphQLSchema } from 'graphql'
import RootQueryType from './rootQueryType'
import mutations from './mutations'

export default new GraphQLSchema({
  query: RootQueryType,
  mutation: mutations,
})
