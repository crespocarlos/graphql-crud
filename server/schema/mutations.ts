import { GraphQLObjectType, GraphQLString, GraphQLID } from 'graphql'
import SongType from './songType'
import LyricType from './lyricType'
import { Song, Lyric } from '../models'

const mutation = new GraphQLObjectType({
  name: 'Mutation',
  fields: {
    addSong: {
      type: SongType,
      args: {
        title: { type: GraphQLString }
      },
      resolve(_, { title }) {
        return new Song({ title }).save()
      }
    },
    addLyricToSong: {
      type: SongType,
      args: {
        content: { type: GraphQLString },
        songId: { type: GraphQLID }
      },
      resolve(_, { content, songId }) {
        return Song.addLyric(songId, content)
      }
    },
    likeLyric: {
      type: LyricType,
      args: { id: { type: GraphQLID } },
      resolve(_, { id }) {
        return Lyric.like(id)
      }
    },
    deleteSong: {
      type: SongType,
      args: { id: { type: GraphQLID } },
      resolve(_, { id }) {
        return Song.remove({ _id: id })
      }
    }
  }
})

export default mutation
