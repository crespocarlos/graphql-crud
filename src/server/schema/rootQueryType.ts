import {
  GraphQLObjectType,
  GraphQLList,
  GraphQLID,
  GraphQLNonNull,
} from 'graphql'
import SongType from './songType'
import LyricType from './lyricType'
import Song from '../models/song'
import Lyric from '../models/lyric'

const RootQuery = new GraphQLObjectType({
  name: 'RootQueryType',
  fields: () => ({
    songs: {
      type: new GraphQLList(SongType),
      resolve() {
        return Song.find({})
      },
    },
    song: {
      type: SongType,
      args: { id: { type: new GraphQLNonNull(GraphQLID) } },
      resolve(_, { id }) {
        return Song.findById(id)
      },
    },
    lyric: {
      type: LyricType,
      args: { id: { type: new GraphQLNonNull(GraphQLID) } },
      resolve(_, { id }) {
        return Lyric.findById(id)
      },
    },
  }),
})

export default RootQuery
