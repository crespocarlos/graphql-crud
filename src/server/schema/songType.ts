import {
  GraphQLObjectType,
  GraphQLString,
  GraphQLID,
  GraphQLList,
} from 'graphql'
import LyricType from './lyricType'
import Song from '../models/song'

const SongType: GraphQLObjectType = new GraphQLObjectType({
  name: 'SongType',
  fields: () => ({
    id: { type: GraphQLID },
    title: { type: GraphQLString },
    lyrics: {
      type: new GraphQLList(LyricType),
      resolve(parentValue) {
        return Song.findLyrics(parentValue.id)
      },
    },
  }),
})

export default SongType
