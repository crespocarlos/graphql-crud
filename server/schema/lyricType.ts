import {
  GraphQLObjectType,
  GraphQLID,
  GraphQLInt,
  GraphQLString
} from 'graphql'
import { Lyric, ILyricModel } from '../models'
import songType from './songType'

const LyricType: GraphQLObjectType = new GraphQLObjectType({
  name: 'LyricType',
  fields: () => ({
    id: { type: GraphQLID },
    likes: { type: GraphQLInt },
    content: { type: GraphQLString },
    song: {
      type: songType,
      resolve(parentValue) {
        return Lyric.findById(parentValue)
          .populate('song')
          .then((lyric: ILyricModel | null) => {
            return lyric ? lyric.song : null
          })
      }
    }
  })
})

export default LyricType
