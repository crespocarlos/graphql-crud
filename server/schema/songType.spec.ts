import { createTestClient } from 'apollo-server-testing'
import schema from './schema'
import { ApolloServer, gql } from 'apollo-server-express'

jest.mock('../models/song', () => ({
  findLyrics: jest.fn(() => [{ id: 2, content: 'bla', likes: 10 }]),
  findById: jest.fn(() => ({ id: 1, title: 'testById' })),
}))

const GET_SONG = gql`
  query SongQuery($id: ID!) {
    song(id: $id) {
      id
      title
      lyrics {
        id
        content
        likes
      }
    }
  }
`

describe('SongType', () => {
  it('fetches single song', async () => {
    // create a test server to test against, using our production typeDefs,
    // resolvers, and dataSources.
    const server = new ApolloServer({
      schema,
    })

    // use the test server to create a query function
    const { query } = createTestClient(server)

    // run query against the server and snapshot the output
    const res = await query({ query: GET_SONG, variables: { id: 1 } })
    expect(res).toMatchSnapshot()
  })
})
