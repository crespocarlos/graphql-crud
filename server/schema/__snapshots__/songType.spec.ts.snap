// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongType fetches single song 1`] = `
Object {
  "data": Object {
    "song": Object {
      "id": "1",
      "lyrics": Array [
        Object {
          "content": "bla",
          "id": "2",
          "likes": 10,
        },
      ],
      "title": "testById",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
